test/functional/mempool_dust.py

```diff
@@ -0,0 +1,112 @@
+    script_to_p2wsh_script,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.test_node import TestNode
+from test_framework.util import (
+    assert_equal,
+    get_fee,
+)
+from test_framework.wallet import MiniWallet
+
+
+DUST_RELAY_TX_FEE = 3000  # default setting [sat/kvB]
+
+
+class DustRelayFeeTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def test_dust_output(self, node: TestNode, dust_relay_fee: Decimal,
+                         output_script: CScript, type_desc: str) -> None:
+        # determine dust threshold (see `GetDustThreshold`)
+        if output_script[0] == OP_RETURN:
+            dust_threshold = 0
+        else:
+            tx_size = len(CTxOut(nValue=0, scriptPubKey=output_script).serialize())
+            tx_size += 67 if output_script.IsWitnessProgram() else 148
+            dust_threshold = int(get_fee(tx_size, dust_relay_fee) * COIN)
+        self.log.info(f"-> Test {type_desc} output (size {len(output_script)}, limit {dust_threshold})")
+
+        # amount right on the dust threshold should pass
+        tx = self.wallet.create_self_transfer()["tx"]
+        tx.vout.append(CTxOut(nValue=dust_threshold, scriptPubKey=output_script))
+        tx_good_hex = tx.serialize().hex()
+        res = node.testmempoolaccept([tx_good_hex])[0]
+        assert_equal(res['allowed'], True)
+
+        # amount just below the dust threshold should fail
+        if dust_threshold > 0:
+            tx.vout[1].nValue -= 1
+            res = node.testmempoolaccept([tx.serialize().hex()])[0]
+            assert_equal(res['allowed'], False)
+            assert_equal(res['reject-reason'], 'dust')
+
+        # finally send the transaction to avoid running out of MiniWallet UTXOs
+        self.wallet.sendrawtransaction(from_node=node, tx_hex=tx_good_hex)
+
+    def run_test(self):
+        self.wallet = MiniWallet(self.nodes[0])
+        self.wallet.rescan_utxos()
+
+        # prepare output scripts of each standard type
+        key = ECKey()
+        key.generate(compressed=False)
+        uncompressed_pubkey = key.get_pubkey().get_bytes()
+        key.generate(compressed=True)
+        pubkey = key.get_pubkey().get_bytes()
+
+        output_scripts = [
+            (key_to_p2pk_script(uncompressed_pubkey),     "P2PK (uncompressed)"),
+            (key_to_p2pk_script(pubkey),                  "P2PK (compressed)"),
+            (key_to_p2pkh_script(pubkey),                 "P2PKH"),
+            (script_to_p2sh_script(CScript([OP_TRUE])),   "P2SH"),
+            (key_to_p2wpkh_script(pubkey),                "P2WPKH"),
+            (script_to_p2wsh_script(CScript([OP_TRUE])),  "P2WSH"),
+            (output_key_to_p2tr_script(pubkey[1:]),       "P2TR"),
+            (program_to_witness_script(2,  b'\x66' * 21), "P2?? (future witness version 2)"),
+            (program_to_witness_script(16, b'\x77' * 40), "P2?? (future witness version 16)"),
```

Comment:
Can you explain the rationale for using 21 bytes and 40 bytes for future witness versions 2 and 16 respectively? Maybe a comment with the standard size for each Script type could be useful for testers and people learning.

Reply:
It could be any size between 2 and 40 bytes. Added a corresponding comment in latest force-push ("# witness programs for segwitv2+ can be between 2 and 40 bytes") and adapted the values accordingly. I prefer now having one with 2 bytes instead of 21 bytes (which was indeed completely arbitrary) as it is the smallest possible output script that is considered standard.